src/main/java/net/onelitefeather/bettergopaint/listeners/InteractListener.java

```diff
@@ -18,97 +18,100 @@
-                } else {
-                    loc = e.getClickedBlock().getLocation().clone();
-                }
-                BukkitAdapter.adapt(p).runAsyncIfFree(() ->epb.getBrush().paint(loc, p, epb));
-            }
+
+        if (event.getAction().isLeftClick() && item.getType().equals(Material.FEATHER)) {
+            PlayerBrush brush = plugin.getBrushManager().getBrush(player);
+            player.openInventory(brush.getInventory());
+            event.setCancelled(true);
+            return;
         }
-        if (e.getPlayer().getItemInHand().getType() == XMaterial.FEATHER.parseMaterial() && (e
-                .getAction()
-                .equals(Action.RIGHT_CLICK_AIR) || e.getAction().equals(Action.RIGHT_CLICK_BLOCK))) {
-            e.setCancelled(true);
-            final Player p = e.getPlayer();
-            final Location loc;
-            if (e.getAction().equals(Action.RIGHT_CLICK_AIR)) {
-                loc = p.getTargetBlock(null, 250).getLocation().clone();
-            } else {
-                loc = e.getClickedBlock().getLocation().clone();
-            }
-            if ((!e.getPlayer().hasPermission("bettergopaint.world.bypass")) && (Settings.settings().GENERIC.DISABLED_WORLDS
-                    .contains(loc.getWorld().getName()))) {
-                return;
-            }
-            if (loc.getBlock().getType().equals(XMaterial.AIR.parseMaterial())) {
-                return;
-            }
-            final PlayerBrush pb = BetterGoPaint.getBrushManager().getPlayerBrush(p);
-
-            if (pb.isEnabled()) {
-                BukkitAdapter.adapt(p).runAsyncIfFree(() -> pb.getBrush().paint(loc, p));
-            } else {
-                p.sendMessage(MiniMessage.miniMessage().deserialize(Settings.settings().GENERIC.PREFIX + "<red>Your brush is " +
-                        "disabled, left click to enable the brush."));
-            }
+
+        if (!event.getAction().isRightClick()) {
+            return;
+        }
+
+        Location location;
+        if (event.getAction().equals(Action.RIGHT_CLICK_AIR)) {
+            Block exact = player.getTargetBlockExact(250, FluidCollisionMode.NEVER);
+            location = exact != null ? exact.getLocation().clone() : null;
+        } else {
+            location = event.getClickedBlock() != null ? event.getClickedBlock().getLocation() : null;
+        }
+
+        if (location == null) {
+            return;
         }
-        if (e.getPlayer().getItemInHand().getType() == XMaterial.FEATHER.parseMaterial() && (e
-                .getAction()
-                .equals(Action.LEFT_CLICK_AIR) || e.getAction().equals(Action.LEFT_CLICK_BLOCK))) {
-            e.setCancelled(true);
-            Player p = e.getPlayer();
-            PlayerBrush pb = BetterGoPaint.getBrushManager().getPlayerBrush(p);
-            p.openInventory(pb.getInventory());
+
+        if ((!player.hasPermission("bettergopaint.world.bypass")) && (Settings.settings().GENERIC.DISABLED_WORLDS
```

Comment:
Also here extract into extra variable

Reply:
again what exactly should i extract? the permission?